client/api.c

```diff
@@ -1904,6 +1904,353 @@ TDNFUpdateInfo(
+                BAIL_ON_TDNF_ERROR(dwError);
+
+                if (qInstalled.count == 0)
+                {
+                    /* We may have found multiples if they occur in multiple
+                       repos. Take the first one. */
+                    queue_push(&qInstall, qResult.elements[0]);
+                }
+                queue_free(&qInstalled);
+            }
+            queue_free(&qResult);
+        }
+        else
+        {
+            dwError = ERROR_TDNF_HISTORY_ERROR;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+    }
+
+    for (int i = 0; i < hd->removed_count; i++)
+    {
+        char *pszPkgName = history_get_nevra(hnm, hd->removed_ids[i]);
+        if (pszPkgName)
+        {
+            dwError = SolvFindSolvablesByNevraStr(pTdnf->pSack->pPool, pszPkgName, &qErase, 1);
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+        else
+        {
+            dwError = ERROR_TDNF_HISTORY_ERROR;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+    }
+
+    dwError = TDNFHistoryGoal(
+                  pTdnf,
+                  &qInstall,
+                  &qErase,
+                  &pSolvedPkgInfo);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    pSolvedPkgInfo->nNeedAction =
+        pSolvedPkgInfo->pPkgsToInstall ||
+        pSolvedPkgInfo->pPkgsToUpgrade ||
+        pSolvedPkgInfo->pPkgsToDowngrade ||
+        pSolvedPkgInfo->pPkgsToRemove  ||
+        pSolvedPkgInfo->pPkgsUnNeeded ||
+        pSolvedPkgInfo->pPkgsToReinstall ||
+        pSolvedPkgInfo->pPkgsObsoleted;
+
+    pSolvedPkgInfo->nNeedDownload =
+        pSolvedPkgInfo->pPkgsToInstall ||
+        pSolvedPkgInfo->pPkgsToUpgrade ||
+        pSolvedPkgInfo->pPkgsToDowngrade ||
+        pSolvedPkgInfo->pPkgsToReinstall;
+
+    pSolvedPkgInfo->ppszPkgsNotResolved = ppszPkgsNotResolved;
+
+    *ppSolvedPkgInfo = pSolvedPkgInfo;
+
+cleanup:
+    history_free_nevra_map(hnm);
+    history_free_delta(hd);
+    destroy_history_ctx(ctx);
+    queue_free(&queueGoal);
+    queue_free(&qInstall);
+    queue_free(&qErase);
+    if (ts) {
+        rpmtsCloseDB(ts);
+        rpmtsFree(ts);
+    }
+    return dwError;
+
+error:
+    if(pSolvedPkgInfo)
+    {
+        TDNFFreeSolvedPackageInfo(pSolvedPkgInfo);
+    }
+    if(ppszPkgsNotResolved)
+    {
+        TDNFFreeStringArray(ppszPkgsNotResolved);
+    }
+    goto cleanup;
+}
+
+uint32_t
+TDNFHistoryList(
+    PTDNF pTdnf,
+    PTDNF_HISTORY_ARGS pHistoryArgs,
+    PTDNF_HISTORY_INFO *ppHistoryInfo)
+{
+    uint32_t dwError = 0;
+    struct history_transaction *tas = NULL;
+    struct history_nevra_map *hnm = NULL;
+    int count = 0;
+    PTDNF_HISTORY_INFO pHistoryInfo = NULL;
+    PTDNF_HISTORY_INFO_ITEM pHistoryInfoItems = NULL;
+    struct history_ctx *ctx = NULL;
+    int rc = 0;
+
+    if(!pTdnf || !pHistoryArgs || !ppHistoryInfo)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (pHistoryArgs->nFrom != 0 && pHistoryArgs->nTo != 0)
```

Comment:
If i am understanding correctly can we directly check for < 0 in first condition itself and one more separate condition for chekcing nFrom > nTo ?

Reply:
Good point, done.